-1

Tested at: 5f3335619fd04a51616162c0a8ba90fe4f262548

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16942/16891/summary.html

I found follow errors while testing this PR

**Failed tests:  ClangBuild**

* **Clang**:

I found a compilation error while trying to compile with clang: 
I used this command: 
 scram b vclean && scram build -k -j 48 USER_CXXFLAGS='-fsyntax-only' COMPILER='llvm compile'


 <pre>>> Compiling edm plugin /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_0_X_2016-12-08-2300/src/IOMC/EventVertexGenerators/src/GaussEvtVtxGenerator.cc 
>> Compiling edm plugin /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_0_X_2016-12-08-2300/src/IOMC/EventVertexGenerators/src/GaussianZBeamSpotFilter.cc 
>> Compiling edm plugin /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_0_X_2016-12-08-2300/src/IOMC/EventVertexGenerators/src/BetafuncEvtVtxGenerator.cc 
>> Compiling edm plugin /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_0_X_2016-12-08-2300/src/IOMC/EventVertexGenerators/test/VtxTester.cc 
>> Compiling edm plugin /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_0_X_2016-12-08-2300/src/IOMC/EventVertexGenerators/test/BoostTester.cc 
/build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_0_X_2016-12-08-2300/src/IOMC/EventVertexGenerators/src/HLLHCEvtVtxGenerator.cc:25:20: error: constexpr variable 'two_pi' must be initialized by a constant expression
  constexpr double two_pi=8.0*std::atan(1.0);
                   ^      ~~~~~~~~~~~~~~~~~~
/build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_0_X_2016-12-08-2300/src/IOMC/EventVertexGenerators/src/HLLHCEvtVtxGenerator.cc:25:31: note: non-constexpr function 'atan' cannot be used in a constant expression
  constexpr double two_pi=8.0*std::atan(1.0);
                              ^
</pre>

--- a/IOMC/EventVertexGenerators/interface/HLLHCEvtVtxGenerator.h
+++ b/IOMC/EventVertexGenerators/interface/HLLHCEvtVtxGenerator.h
@@ -99,14 +99,11 @@ private:
     // crossing angle * crab frequency
     const double phiCR;
     
-    //width for x/y plane
+    //width for y plane
     double sigma(double z, double epsilon, double beta, double betagamma) const;
 
-    //width for x/y plane
-    double rhozt(double z, double t) const;
-        
     //density with crabbing
-    double integrandCC(double z, double t) const;
+    double integrandCC(double x, double z, double t) const;
 
     // 4D intensity
     double intensity(double x, double y,double z,double t) const;

--- a/IOMC/EventVertexGenerators/python/VtxSmearedParameters_cfi.py
+++ b/IOMC/EventVertexGenerators/python/VtxSmearedParameters_cfi.py
@@ -638,7 +638,7 @@ HLLHCVtxSmearingParameters = cms.PSet(
     BetaSeparationPlaneInm = cms.double(0.20),
     HorizontalEmittance = cms.double(2.5e-06),
     VerticalEmittance = cms.double(2.05e-06),
-    BunchLengthInm = cms.double(0.09),
+    BunchLengthInm = cms.double(0.090),
     CrabbingAngleCrossingInurad = cms.double(380.0),
     CrabbingAngleSeparationInurad = cms.double(0.0)    
 )

--- a/IOMC/EventVertexGenerators/src/HLLHCEvtVtxGenerator.cc
+++ b/IOMC/EventVertexGenerators/src/HLLHCEvtVtxGenerator.cc
@@ -22,6 +22,8 @@ namespace {
   constexpr double gamma34 = 1.22541670246517764513;    // Gamma(3/4)
   constexpr double gamma14 = 3.62560990822190831193;    // Gamma(1/4)
   constexpr double gamma54 = 0.90640247705547798267;    // Gamma(5/4)
+  constexpr double sqrt2 = 1.41421356237;
+  constexpr double sqrt2to5 = 5.65685424949;
   constexpr double two_pi=2.0*M_PI;
 }
 
@@ -200,8 +202,8 @@ double HLLHCEvtVtxGenerator::integrandCC(double z,
     
   const double sigs2 = sigs*sigs;
   
-  constexpr double factorRMSgauss4  = 1./std::sqrt(2.)/gamma34 * gamma14; // # Factor to take rms sigma as input of the supergaussian
-  constexpr double NormFactorGauss4 = std::pow(2.0,(5./2.)) * gamma54 * gamma54;
+  constexpr double factorRMSgauss4  = 1./sqrt2/gamma34 * gamma14; // # Factor to take rms sigma as input of the supergaussian
+  constexpr double NormFactorGauss4 = sqrt2to5 * gamma54 * gamma54;
   
   const double sinCR  = std::sin(phiCR/2.0);
   const double sinCR2 = sinCR*sinCR;

--- a/IOMC/EventVertexGenerators/src/HLLHCEvtVtxGenerator.cc
+++ b/IOMC/EventVertexGenerators/src/HLLHCEvtVtxGenerator.cc
@@ -22,7 +22,7 @@ namespace {
   constexpr double gamma34 = 1.22541670246517764513;    // Gamma(3/4)
   constexpr double gamma14 = 3.62560990822190831193;    // Gamma(1/4)
   constexpr double gamma54 = 0.90640247705547798267;    // Gamma(5/4)
-  constexpr double two_pi=8.0*std::atan(1.0);
+  constexpr double two_pi=2.0*M_PI;
 }
 
 void HLLHCEvtVtxGenerator::fillDescriptions(edm::ConfigurationDescriptions &descriptions)

--- a/IOMC/EventVertexGenerators/src/HLLHCEvtVtxGenerator.cc
+++ b/IOMC/EventVertexGenerators/src/HLLHCEvtVtxGenerator.cc
@@ -97,7 +97,7 @@ HepMC::FourVector* HLLHCEvtVtxGenerator::newVertex(CLHEP::HepRandomEngine* engin
     z=(shoot()-0.5)*6.0*sigs;
     t=(shoot()-0.5)*6.0*sigs;
     x=(shoot()-0.5)*12.0*sigma(0.0,epsxn,betx,betagamma);
-    y=(shoot()-0.5)*8.0*sigma(0.0,epssn,bets,betagamma);
+    y=(shoot()-0.5)*12.0*sigma(0.0,epssn,bets,betagamma);
     
     i=intensity(x,y,z,t);
     
@@ -137,12 +137,6 @@ double HLLHCEvtVtxGenerator::sigma(double z, double epsilon, double beta, double
   return sigma;
 } 
 
-double HLLHCEvtVtxGenerator::rhozt(double z, double ct) const 
-{  
-  return integrandCC(z,ct); 
-}
-
-
 double HLLHCEvtVtxGenerator::intensity(double x, 
                                        double y, 
                                        double z, 
@@ -150,41 +144,25 @@ double HLLHCEvtVtxGenerator::intensity(double x,
   //---c in m/s --- remember t is already in meters
   constexpr double c= 2.99792458e+8; // m/s
   
-  const double sigmax=sigma(z,epsxn,betx,betagamma);
   const double sigmay=sigma(z,epssn,bets,betagamma);
     
-  const double alphax_mod=alphax;//*std::cos(wcc*(z-t)/c);
-  const double alphay_mod=alphay;//*std::cos(wcc*(z-t)/c);
+  const double alphay_mod=alphay*std::cos(wcc*(z-t)/c);
   
-  const double cax       = std::cos(alphax_mod);
-  const double sax_minus = std::sin(-alphax_mod);
-  const double sax_plus  = -sax_minus;
-
   const double cay=std::cos(alphay_mod);
   const double say=std::sin(alphay_mod);
 
-  const double ct       = std::cos(0.5*phi);
-  const double st_minus = std::sin(-0.5*phi);
-  const double st_plus  = -st_minus;
+  const double dy=-(z-t)*say-y*cay;
 
-  const double dx_plus  = (z-t*ct)*(cax*st_plus-sax_plus*ct)   - (x-t*st_plus)*(sax_plus*st_plus+cax*ct);
-  const double dx_minus = (z+t*ct)*(cax*st_minus-sax_minus*ct) - (x+t*st_minus)*(sax_minus*st_minus+cax*ct);
+  const double xzt_density= integrandCC(x,z,t);
 
-  const double dx=/*-(z-t)*sax_plus*/ -x*cax; 
-
-  const double dy=/*-(z-t)*say*/-y*cay;
-
-  const double zrho=rhozt(z,t);
-
-  const double norm = (two_pi*sigmax*sigmay);
+  const double norm = two_pi*sigmay;
   
-  return ( std::exp(-0.5*dx*dx/(sigmax*sigmax))*
-           std::exp(-0.5*dx*dx/(sigmax*sigmax))*
-           std::exp(-dy*dy/(sigmay*sigmay))*
-           zrho/(norm*norm) );
+  return ( std::exp(-dy*dy/(sigmay*sigmay))*
+           xzt_density/norm );
 }
 
-double HLLHCEvtVtxGenerator::integrandCC(double z, 
+double HLLHCEvtVtxGenerator::integrandCC(double x,
+                                         double z, 
                                          double ct) const {  
   constexpr double local_c_light = 2.99792458e8;
 
@@ -215,16 +193,19 @@ double HLLHCEvtVtxGenerator::integrandCC(double z,
     result = norm*std::exp(-ct*ct/sigs2
                            -z*z*cos2/sigs2
                            -1.0/(4*k2*sigmax2)*(
-                                                -4*cosks*cosks * sinkct*sinkct * sinCR2
+                                                //-4*cosks*cosks * sinkct*sinkct * sinCR2 // comes from integral over x
                                                 -8*z*k*std::sin(k*z)*std::cos(k*ct) * sin * sinCR
                                                 +2 * sinCR2
                                                 -std::cos(2*k*(z-ct)) * sinCR2
                                                 -std::cos(2*k*(z+ct)) * sinCR2
                                                 +4*k2*z*z *sin2
                                                 )
+                           - x*x*(cos2/sigmax2 + sin2/sigs2) // contribution from x integrand
+                           + x*ct*sin/sigs2 // contribution from x integrand
+                           + 2*x*cos*cosks*sinkct*sinCR/k/sigmax2 // contribution from x integrand
                            //+(2*ct/k)*np.cos(k*s)*np.sin(k*ct) *(sin*sinCR)/(sigs2*cos)  # small term
                            //+ct**2*(sin2/sigs4)/(cos2/sigmax2)				              # small term
-                           )/std::sqrt(1.0+(z*z)/(betx*betx))/std::sqrt(1.0+(z*z)/(bets*bets));
+                           )/(1.0+(z*z)/(betx*betx))/std::sqrt(1.0+(z*z)/(bets*bets));
       
   } else {
     

